clients/src/main/java/org/apache/kafka/clients/admin/ListConsumerGroupsResult.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients.admin;
+
+import org.apache.kafka.common.KafkaFuture;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.annotation.InterfaceStability;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The result of the {@link AdminClient#listConsumerGroups()} call.
+ * <p>
+ * The API of this class is evolving, see {@link AdminClient} for details.
+ */
+@InterfaceStability.Evolving
+public class ListConsumerGroupsResult {
+    final KafkaFuture<Map<Node, KafkaFuture<Collection<ConsumerGroupListing>>>> futureMap;
+
+    ListConsumerGroupsResult(KafkaFuture<Map<Node, KafkaFuture<Collection<ConsumerGroupListing>>>> futureMap) {
```

Comment:
Agree, but I was wondering: if a Node is not available, we could return a list with less consumer groups, right? In that case, a map with Node -> List[ConsumerGroupListing] could help to compare results against a list of Nodes.

If this is not an issue, I was thinking about changing the result to: `KafkaFuture<Collection<KafkaFuture<ConsumerGroupListing>>>` 
make sense?

Reply:
That is a good point.

I think adding a boolean flag in addition to the `future` result indicating if it may be incomplete is better than exposing the node map. I.e. we could have two fields inside `ListConsumerGroupsResult`, a `boolean` and a `KafkaFuture<Collection<ConsumerGroupListing>>` (I think we do not need to have nested KafkaFutures so I did not have that inside `Collection`, but correct me if I overlook anything).

Reply:
Agree. Would be OK to get list of nodes from `describeCluster` and then join futures?:

```
    private Collection<Node> listNodes(Integer timeoutMs) {
        try {
            return describeCluster(new DescribeClusterOptions().timeoutMs(timeoutMs)).nodes().get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }
```

```
        final Collection<Node> nodes = listNodes(options.timeoutMs);

        final Set<KafkaFutureImpl<Collection<ConsumerGroupListing>>> futures = new HashSet<>(nodes.size());

        for (final Node node : nodes) {
          ...
        }

        KafkaFuture.allOf(futures.toArray(new KafkaFuture[0]))
                .thenApply(new KafkaFuture.Function<Void, Collection<ConsumerGroupListing>>() {
                    @Override
                    public Collection<ConsumerGroupListing> apply(Void aVoid) {
                        final Set<ConsumerGroupListing> consumerGroups = new HashSet<>();
                        for (KafkaFutureImpl<Collection<ConsumerGroupListing>> future : futures) {
                            try {
                                consumerGroups.addAll(future.get());
                            } catch (InterruptedException | ExecutionException e) {
                                // This should be unreachable, because allOf ensured that all the futures
                                // completed successfully.
                                throw new RuntimeException(e);
                            }
                        }
                        return consumerGroups;
                    }
                });
```